python/GafferImageTest/ImageWriterTest.py

```diff
@@ -1787,6 +1788,113 @@ def testWithEXRSampleImages( self ):
 					rereader["channelInterpretation"].setValue( GafferImage.ImageReader.ChannelInterpretation.Specification )
 					self.assertImagesEqual( rereader["out"], reader["out"], ignoreMetadata = True, ignoreChannelNamesOrder = ignoreOrder, ignoreDataWindow = expandDataWindow )
 
+	def channelTypesFromHeader( self, path ):
+		r = subprocess.check_output( ["exrheader", path ], universal_newlines=True )
+
+		channelText = re.findall( re.compile( r'channels \(type chlist\):\n((    .*\n)+)', re.MULTILINE ), r )
+
+		return [ i.split()[:2] for part in channelText for i in part[0].splitlines()  ]
+
+	def testPerChannelDataType( self ):
+
+		reference = self.channelTestImage()
+
+		writer = GafferImage.ImageWriter()
+		writer["in"].setInput( reference["out"] )
+		writer["fileName"].setValue( os.path.join( self.temporaryDirectory(), "test.tif" ) )
+
+		# Start with some formats that don't support per channel data types, and check that they fail as
+		# expected
+		writer["dataTypeExpression"] = Gaffer.Expression()
+		writer["dataTypeExpression"].setExpression( 'parent["tiff"]["dataType"] = "uint8" if context.get( "imageWriter:channelName" ) == "R" else "uint16"' )
+
+		with six.assertRaisesRegex( self, Gaffer.ProcessException, "File format tiff does not support per-channel data types" ) :
+			writer["task"].execute()
+
+		writer["fileName"].setValue( os.path.join( self.temporaryDirectory(), "test.dpx" ) )
+		writer["dataTypeExpression"].setExpression( 'parent["dpx"]["dataType"] = "uint8" if context.get( "imageWriter:channelName" ) == "R" else "uint16"' )
+
+		with six.assertRaisesRegex( self, Gaffer.ProcessException, "File format dpx does not support per-channel data types" ) :
+			writer["task"].execute()
+
+		writePath = os.path.join( self.temporaryDirectory(), "test.exr" )
+		writer["fileName"].setValue( writePath )
+
+		writer["task"].execute()
+
+		# Default channel types - depthDataType forces Z and ZBack to 32-bit, but not character.Z
```

Comment:
I know this is the existing behaviour, but why would `character.Z` not qualify as depth data?

Reply:
As per conversation this morning:  I think there is an argument that this behaviour is reasonable - it generally doesn't make sense to store more than one depth per pixel, so "character.Z" is a pretty artificial example.  If Z is somehow representing something other than the global distance from camera ( maybe fog depth or something? ), then it quite possibly doesn't need the extra precision.

But anyway, we agreed that it doesn't need to be changed as part of this PR, and we can just keep backwards compatibility.  If someone wants something more general, they can turn off depthDataType and use an expression on dataType now.